semantic/annotate.go

```diff
@@ -0,0 +1,67 @@
+package semantic
+
+import (
+	"strconv"
+)
+
+type typeVar struct {
+	prefix string
+	suffix int
+}
+
+// TypeAnnotationVisitor visits the nodes in a semantic
+// graph and assigns all sub-expressions type variables.
+type TypeAnnotationVisitor struct {
+	// tenv represents the type environment of a program.
+	// It maps a semantic node to a unique type variable.
+	tenv map[Node]Type
+	// the next type variable to be assigned. The first
+	// type variable is t0. Then t1, t2, t3 and so on.
+	next *typeVar
+}
+
+// Visit assigns a new type variable to a semantic node
+func (v *TypeAnnotationVisitor) Visit(node Node) Visitor {
+	if n, ok := node.(Expression); ok {
+		v.tenv[n] = v.newTypeVar()
```

Comment:
Should this return here?

Reply:
No, it could be a call expression in which case it needs to perform a pipe check.

Reply:
Never mind. Yes by changing `semantic.Walk` to also visit the pipe expression if it exists, we may just return after this if statement.